src/utils/theme/theme.provider.tsx

```diff
@@ -0,0 +1,249 @@
+  }, [themeMode, getThemeState]);
+
+  // (ONLY CONNECT, NOT BE) Add `data-theme` attribute based on resolved theme (light or dark, not system):
+
+  useEffect(() => {
+    if (import.meta.env?.VITE_IS_EMBEDDED_APP !== "1") return;
+
+    document.documentElement.setAttribute("data-theme", displayTheme);
+  }, [displayTheme]);
+
+  // (ONLY CONNECT, NOT BE) Emit "THEME_CHANGE" event:
+
+  useEffect(() => {
+    if (import.meta.env?.VITE_IS_EMBEDDED_APP !== "1" || !isInsideIframe()) return;
+
+    if (window.parent !== window) {
+      // TODO: Update to use postEmbeddedMessage:
+      window.parent.postMessage({ type: "THEME_CHANGE", data: themeState }, "*");
+    }
+  }, [themeState]);
+
+  // Reduced motion preference (for Framer Motion):
+
+  useEffect(() => {
+    const reducedMotionPreference = window.matchMedia("(prefers-reduced-motion)");
+
+    if (reducedMotionPreference.matches) {
+      // This could also be always set to `true` at the top of the file and then set to `false` after the application
+      // loads if we still notice some flicker due to any animation/transition playing when the view first loads.
+
+      MotionGlobalConfig.skipAnimations = true;
+    }
+  }, []);
+
+  // Focus active global class:
+
+  useEffect(() => {
+    function handleKeyDown({ code }: KeyboardEvent) {
+      switch (code) {
+        case "Tab": {
+          document.documentElement.classList.add(IS_FOCUS_ACTIVE_CLS);
+          break;
+        }
+
+        case "Escape": {
+          const hadFocusActive = document.documentElement.classList.contains(IS_FOCUS_ACTIVE_CLS);
+
+          document.documentElement.classList.remove(IS_FOCUS_ACTIVE_CLS);
+
+          if (
+            import.meta.env?.VITE_IS_EMBEDDED_APP === "1" &&
+            (!hadFocusActive || document.activeElement === document.documentElement)
+          ) {
+            postEmbeddedMessage({
+              type: "embedded_close",
+              data: null,
+            });
+          }
+
+          break;
+        }
+      }
+    }
+
+    function handleMouseDown({ target }: MouseEvent) {
+      if (document.activeElement !== target) document.documentElement.classList.remove(IS_FOCUS_ACTIVE_CLS);
+    }
+
+    document.addEventListener("keydown", handleKeyDown);
+    document.addEventListener("mousedown", handleMouseDown);
+
+    return () => {
+      document.removeEventListener("keydown", handleKeyDown);
+      document.removeEventListener("mousedown", handleMouseDown);
+    };
+  }, []);
+
+  // Memoize the context value for better performance:
+
+  const themeContextValue: ThemeContextValue = useMemo(
+    () => ({
+      ...themeState,
+      isDarkMode: themeState.displayTheme === "dark",
+      tokens: themeTokens,
+      common: commonTokens,
+      setTheme,
+    }),
+    [themeState, setTheme],
+  );
+
+  return (
+    <ThemeContext.Provider value={themeContextValue}>
+      <ThemeBackgroundObserver displayTheme={displayTheme} />
```

Comment:
`ThemeBackgroundObserver` should be moved inside `StyledComponentsThemeProvider` as it uses `useTheme` hook and theme is `undefined` unless it's inside `StyledComponentsThemeProvider` context.